Firmware/HomeButtonsArduino/src/mdi_helper.cpp

```diff
@@ -0,0 +1,205 @@
+}
+
+bool MDIHelper::download(const char* name, uint16_t size) {
+  if (!spiffs_mounted_) {
+    error("SPIFFS not mounted");
+    return false;
+  }
+
+  char path[64];
+  _get_path(path, name, size);
+
+  if (SPIFFS.exists(path)) {
+    info("'%s' already exists", name);
+    return true;
+  }
+
+  debug("Downloading '%s' to '%s'", name, path);
+
+  File file = SPIFFS.open(path, FILE_WRITE, true);
+  if (!file) {
+    error("Failed to open '%s' for writing", path);
+    return false;
+  }
+
+  StaticString<256> url("%s%dx%d/%s.bmp", MDI_URL, size, size, name);
+  bool ret = download::download_file_https(
+      HOST, url.c_str(), file,
+      github_raw_cert::cert_DigiCert_TLS_RSA_SHA256_2020_CA1);
+  if (ret) {
+    info("Downloaded '%s'", name);
+    return true;
+  } else {
+    error("Failed to download '%s'", name);
+    SPIFFS.remove(path);
+    return false;
+  }
+}
+
+bool MDIHelper::download(const char* name) {
+  if (!spiffs_mounted_) {
+    error("SPIFFS not mounted");
+    return false;
+  }
+
+  for (uint8_t i = 0; i < num_sizes_; ++i) {
+    if (!download(name, sizes_[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool MDIHelper::exists(const char* name, uint16_t size) {
+  if (!spiffs_mounted_) {
+    error("SPIFFS not mounted");
+    return false;
+  }
+  char path[64];
+  _get_path(path, name, size);
+  return SPIFFS.exists(path);
+}
+
+bool MDIHelper::exists(const char* name) {
+  if (!spiffs_mounted_) {
+    error("SPIFFS not mounted");
+    return false;
+  }
+  for (uint8_t i = 0; i < num_sizes_; ++i) {
+    if (!exists(name, sizes_[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+File MDIHelper::get_file(const char* name, uint16_t size) {
+  if (!spiffs_mounted_) {
+    error("SPIFFS not mounted");
+    return File();
+  }
+
+  if (!exists(name, size)) {
+    error("'%s' does not exist", name);
+    return File();
+  }
+
+  char path[64];
+  _get_path(path, name, size);
+  debug("Opening '%s'", path);
+  return SPIFFS.open(path, FILE_READ);
+}
+
+size_t MDIHelper::get_free_space() {
+  if (!spiffs_mounted_) {
+    error("SPIFFS not mounted");
+    return 0;
+  }
+  size_t free = SPIFFS.totalBytes() - SPIFFS.usedBytes();
+  debug("Free space: %d", free);
+  return free;
+}
+
+bool MDIHelper::make_space(size_t size) {
+  if (!spiffs_mounted_) {
+    error("SPIFFS not mounted");
+    return false;
+  }
+  if (get_free_space() > size) {
+    return true;
+  }
+  info("Freeing space...");
+  File root = SPIFFS.open(FOLDER);
+  if (!root) {
+    error("Failed to open '%s'", FOLDER);
+    return false;
+  }
+  if (!root.isDirectory()) {
+    error("'%s' is not a directory", FOLDER);
+    return false;
+  }
+  uint16_t count = 0;
+  size_t size_before = SPIFFS.usedBytes();
+  while (get_free_space() < size) {
+    File file = root.openNextFile();
+    if (!file) {
+      error("Failed to open next file");
+      continue;
+    }
```

Comment:
What is preventing this code from looping indefinitely ?
When last file is reached, what is `File file = root.openNextFile();` really returning ?
In case the file isn't valid, you simply skip the iteration (`continue`).